resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala

```diff
@@ -381,4 +382,13 @@ object KubernetesUtils extends Logging {
       }
     }
   }
+
+  // Load and init custom feature step according to `className` and `conf`
+  // This method is used by `KubernetesDriverBuilder` and `KubernetesExecutorBuilder`
+  @Since("3.3.0")
+  def loadFeatureStep(conf: KubernetesConf, className: String): KubernetesFeatureConfigStep = {
```

Comment:
I would remove this `def` and would use something like like this instead in for example at `KubernetesExecutorBuilder`:
```scala
        val feature = Utils.classForName[Any](className).newInstance()
        val initializedFeature = feature match {
          case e: KubernetesExecutorCustomFeatureConfigStep =>
            e.init(conf)
            Some(e)
          case _: KubernetesDriverCustomFeatureConfigStep =>
            None
          case b: KubernetesCustomFeatureConfigStep => 
            Some(b)
          case _ =>
            None
        }
        initializedFeature.getOrElse {
          logError("...")
          throw new SparkException("...")
        }
```

This way `conf` has the correct type (`KubernetesExecutorConf`) and we can check for the correct accpeted types. Regarding  `KubernetesDriverCustomFeatureConfigStep` I won't introduce the `init` at all.

If you agree please run some manual tests with wrong types for both for the executor and for the driver.

Reply:
> Regarding `KubernetesDriverCustomFeatureConfigStep` I won't introduce the init at all.

`KubernetesDriverCustomFeatureConfigStep` is used for a custom feature step  with base KubernetesConf, can be shared by driver and executor. and previous `KubernetesFeatureConfigStep` can without conf.

So I guess then final look is like below:
```scala
        val initializedFeature = feature match {
          // Since 3.3, allow user implements feature with KubernetesExecutorConf
          case e: KubernetesExecutorCustomFeatureConfigStep =>
            e.init(conf)
            Some(e)
          // raise SparkException with wrong type feature step
          case _: KubernetesDriverCustomFeatureConfigStep =>
            None
          // Since 3.3, allow user implements feature with KubernetesConf
          case c: KubernetesCustomFeatureConfigStep =>
            c.init(conf)
            Some(c)
          // Since 3.2, allow user implements feature without config
          case f: KubernetesFeatureConfigStep =>
            Some(f)
          case _ => None
        }
```

WDYT?

Reply:
Fine for me.

Reply:
Although one can do the following even without `KubernetesCustomFeatureConfigStep#init`: 

```scala
class MyStep extends  KubernetesDriverCustomFeatureConfigStep with KubernetesExecutorCustomFeatureConfigStep {

  def init(conf: KubernetesDriverConf): Unit {}

  def init(conf: KubernetesExecutorConf): Unit {}

}
```

Reply:
Ah, yes, but it looks like a little bit wired.

I have updated the PR, you could take a look when you have time. Thanks.

Reply:
Ah, yes, I got your idea completely. I think it's also ok for me.

Reply:
@attilapiros I have updated the PR, test and desciption as your  suggestion, it would be good if you can get time to review again. : ) Thanks!